### PR TITLE
feat: add dynamodb-local module

### PR DESCRIPTION
**Pull request by mdelapenya**
## What does this PR do?
This PR copies the already existing community module for DynamoDB (see https://github.com/abhirockzz/dynamodb-local-testcontainers-go), using most recent APIs and patterns.

## Why is it important?
The community has asked for updates in the community module without response. We'd like to enable them in using this techonology.

## Related issues

- Relates https://github.com/abhirockzz/dynamodb-local-testcontainers-go/pull/3
- Relates https://github.com/abhirockzz/dynamodb-local-testcontainers-go/pull/5
- Relates https://github.com/abhirockzz/dynamodb-local-testcontainers-go/pull/6
- Relates https://github.com/abhirockzz/dynamodb-local-testcontainers-go/pull/7

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -129,7 +129,7 @@ jobs:
     runs-on: ubuntu-latest 
     steps:
       - name: Check out code into the Go module directory
-        uses: actions/checkout@692973e3d937129bcbf40652eb9f2f61becf3332 # v4.1.7
+        uses: actions/checkout@692973e3d937129bcbf40652eb9f2f61becf3332 # v4.1.1
```
Comment (mdelapenya) [inline .github/workflows/ci.yml:132]: Added by the modulegen
```suggestion
        uses: actions/checkout@692973e3d937129bcbf40652eb9f2f61becf3332 # v4.1.7
```

File: modules/dynamodb/dynamodb.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+	if err != nil {
+		return c, fmt.Errorf("generic container: %w", err)
+	}
+
+	return c, nil
+}
+
+// ConnectionString returns DynamoDB local endpoint host and port in <host>:<port> format
+func (c *DynamoDBContainer) ConnectionString(ctx context.Context) (string, error) {
+	var errs []error
```
Comment (stevenh) [inline modules/dynamodb/dynamodb.go:59]: suggestion: we just want the wrapping here not best effort.

File: modules/dynamodb/dynamodb.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/testcontainers/testcontainers-go"
+	"github.com/testcontainers/testcontainers-go/wait"
+)
+
+const (
+	image         = "amazon/dynamodb-local:2.2.1"
```
Comment (stevenh) [inline modules/dynamodb/dynamodb.go:13]: suggestion: this looks unused
```suggestion
```

File: modules/dynamodb/dynamodb_test.go [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+		AttributeDefinitions: []types.AttributeDefinition{
+			{
+				AttributeName: aws.String(pkColumnName),
+				AttributeType: types.ScalarAttributeTypeS,
+			},
+		},
+		BillingMode: types.BillingModePayPerRequest,
+	})
+	if err != nil {
+		return err
```
Comment (stevenh) [inline modules/dynamodb/dynamodb_test.go:179]: suggestion: wrap
```suggestion
		return fmt.Errorf("create table %q: %w", tableName, err)
```

File: modules/dynamodb/dynamodb_test.go [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+
+func queryItem(client *dynamodb.Client, val string) (string, error) {
+	output, err := client.GetItem(context.Background(), &dynamodb.GetItemInput{
+		TableName: aws.String(tableName),
+		Key: map[string]types.AttributeValue{
+			pkColumnName: &types.AttributeValueMemberS{Value: val},
+		},
+	})
+	if err != nil {
+		return "", err
```
Comment (stevenh) [inline modules/dynamodb/dynamodb_test.go:207]: suggestion: wrap
```suggestion
		return fmt.Errorf("get item %q: %w", val, err)
```

File: modules/dynamodb/dynamodb_test.go [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+}
+
+func (r *dynamoDBResolver) ResolveEndpoint(ctx context.Context, params dynamodb.EndpointParameters) (smithyendpoints.Endpoint, error) {
+	return smithyendpoints.Endpoint{
+		URI: url.URL{Host: r.HostPort, Scheme: "http"},
+	}, nil
+}
+
+// getDynamoDBClient returns a new DynamoDB client with the endpoint resolver set to the DynamoDB container's host and port
+func getDynamoDBClient(c *tcdynamodb.DynamoDBContainer) (*dynamodb.Client, error) {
```
Comment (stevenh) [inline modules/dynamodb/dynamodb_test.go:226]: suggestion: this is returning an error only to check its not one, pass in t and use require to simplify.

File: modules/dynamodb/dynamodb_test.go [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func queryItem(client *dynamodb.Client, val string) (string, error) {
```
Comment (stevenh) [inline modules/dynamodb/dynamodb_test.go:199]: suggestion: this is returning an error only to require its not one, pass in t to simplify.

File: modules/dynamodb/dynamodb_test.go [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+		BillingMode: types.BillingModePayPerRequest,
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func addDataToTable(client *dynamodb.Client, val string) error {
```
Comment (stevenh) [inline modules/dynamodb/dynamodb_test.go:185]: suggestion: this is returning an error only to assert its not one, pass in t to simplify.

File: modules/dynamodb/dynamodb_test.go [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+		URI: url.URL{Host: r.HostPort, Scheme: "http"},
+	}, nil
+}
+
+// getDynamoDBClient returns a new DynamoDB client with the endpoint resolver set to the DynamoDB container's host and port
+func getDynamoDBClient(c *tcdynamodb.DynamoDBContainer) (*dynamodb.Client, error) {
+	// createClient {
+	hostPort, err := c.ConnectionString(context.Background())
+	if err != nil {
+		return nil, err
```
Comment (stevenh) [inline modules/dynamodb/dynamodb_test.go:230]: suggestion: wrap errors, more below

**Review by stevenh — CHANGES_REQUESTED** [review]

File: modules/dynamodb/dynamodb_test.go (hunk 7 above)
Comment (mdelapenya) [reply modules/dynamodb/dynamodb_test.go:193]: There are cases where we check for error and for no error, so I think it's fine to keep it like it is now

File: modules/dynamodb/dynamodb.go [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+		return "", err
+	}
+
+	hostIP, err := c.Host(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	uri := fmt.Sprintf("%s:%s", hostIP, mappedPort.Port())
+	return uri, nil
```
Comment (stevenh) [inline modules/dynamodb/dynamodb.go:68]: nit: we could avoid the variable and sprintf.
```suggestion
	return hostIP+":"+mappedPort.Port(), nil
```

**Review by stevenh — DISMISSED** [review]

**Review by stevenh — APPROVED** [review]

**Pull request merged**